### DISCUSSION TITLE
How to make replacement values?

### DISCUSSION THREAD

**Discussion by zombiebober** [cue-lang/cue#1354] [Q&A] [answered]
Hello. My project has a file (vars.yaml) that stores all the data for the schemas.
But sometimes, you need to generate a configuration taking into account another configuration file that stores part of the data in it.

It is necessary to replace certain data in an already completed schema. How can I do that?

main.cue

```cue

package template

// schemas
#Global: {
	host: string
}

#Mode: {
	user: string
	license: email: string
}

// yaml vars validation
global: #Global
mode: [env=string]: #Mode

#env: string | *"dev" @tag(env)

#Output: {
	#IN: {
		global: #Global
		vars:   #Mode
	}

	result: {
		service: {
			host:  #IN.global.host
			user:  "\(#IN.vars.user)@\(#IN.global.host)"
			email: #IN.vars.license.email
		}
	}
}

final: (#Output & {#IN: {"global": global, vars: mode[#env]}}).result

```

vars.yaml
```yaml
global:
  host: domain.com


mode:
  dev:
    user: "dev"
    license:
      email: admin@dev.com
  prd:
    user: "prd"
    license:
      email: admin@prd.com
```

When executing this command, the result will be as follows.
```shell
cue eval main.cue vars.yaml --out yaml -e final -t env=prd
service:
  host: domain.com
  user: prd@domain.com
  email: admin@prd.com
```

I want to change the data in the schema

general.cue

```cue
package template

general: {
	host?:  string
	user?:  string
	email?: string
}

if general.host != _|_ {
	global.host: general.host
}
if general.user != _|_ {
	var.user: general.user
}
if general.email != _|_ {
	var.license.email: general.email
}
```
general.yaml
```yaml
host: custom.net
email: user@custom.com

```

When executing this command, the result will be as follows.
```shell
cue eval main.cue vars.yaml general.yaml --out yaml -e final -t env=prd
service:
  host: custom.net
  user: prd@custom.net
  email: user@custom.com
```
```shell
cue eval main.cue vars.yaml general.yaml --out yaml -e final 
service:
  host: custom.net
  user: dev@dcustom.net
  email: user@custom.com
```

**Comment by verdverm** [comment, marked as answer]
Generally, you cannot replace values. Here are a few options to consider

- Needing to replace is a sign of an anti-pattern. You will want to rethink how your CUE and configuration is organized.
- Best practice would have you arrange defaults so that they work out as data are resolved. Maybe something like `actual: *general.field | ( *global.field | string)`. Note the nesting of default marks.
- Using `general` as a default like `#Global: host: string | *general.host`. I'm not sure if this would work for your `email` example as there are already concrete values there.
- Checking both sources and producing an intermediate value per field. Where and how you do this will have maintainability trade-offs. You might make a "function" in `general.cue` and use it in `#Output`. It would likely take `#In` as input and produce a `#Mid: (#General & { #in: #In }).result`. You would then reference `#Mid` instead of `#In`.
- I've been experimenting with a general purpose replacer. You would still need to have a new defn to reference in `#Output`. Maybe it can be helpful to use or look at for ideas: [#Replace](https://github.com/hofstadter-io/cuetils/blob/main/structural/replace.cue) and [Example](https://github.com/hofstadter-io/cuetils/blob/main/test/cue/replace.cue#L37).
